src/cache/cache-manager.ts

```diff
@@ -0,0 +1,152 @@
+      const keyManifestEntry = await this.keyManifest.get();
+
+      if (keyManifestEntry) {
+        const matchedKey = this.matchExistingCacheKey(
+          cacheKey,
+          keyManifestEntry.keys
+        );
+        wrappedEntry = await this.cache.get<WrappedCacheEntry>(matchedKey);
+      }
+
+      // If we still don't have an entry..
+      if (!wrappedEntry) {
+        return;
+      }
+    }
+
+    const nowSeconds = Math.floor(Date.now() / 1000);
+
+    // Make sure the key manifest knows about the key.
+    // This helps to migrate keys into the manifest, as the manifest takes care
+    // of duplicates for us.
+    await this.keyManifest.add(cacheKey.toKey());
+
+    if (wrappedEntry.expiresAt - expiryAdjustmentSeconds < nowSeconds) {
+      if (wrappedEntry.body.refresh_token) {
+        wrappedEntry.body = {
+          refresh_token: wrappedEntry.body.refresh_token
+        };
+
+        await this.cache.set(cacheKey.toKey(), wrappedEntry);
+        return wrappedEntry.body;
+      }
+
+      await this.cache.remove(cacheKey.toKey());
+      await this.keyManifest.remove(cacheKey.toKey());
+      return;
+    }
+
+    return wrappedEntry.body;
+  }
+
+  async set(entry: CacheEntry): Promise<void> {
+    const cacheKey = new CacheKey({
+      client_id: entry.client_id,
+      scope: entry.scope,
+      audience: entry.audience
+    });
+
+    const wrappedEntry = this.wrapCacheEntry(entry);
+
+    await this.cache.set(cacheKey.toKey(), wrappedEntry);
+    await this.keyManifest.add(cacheKey.toKey());
+  }
+
+  async clear(): Promise<void> {
+    const keyManifest = await this.keyManifest.get();
+
+    if (keyManifest) {
+      keyManifest.keys.forEach(async key => {
+        await this.cache.remove(key);
+      });
+
+      await this.keyManifest.clear();
+    }
+  }
+
+  private wrapCacheEntry(entry: CacheEntry): WrappedCacheEntry {
+    const expiresInTime = Math.floor(Date.now() / 1000) + entry.expires_in;
+
+    const expirySeconds = Math.min(
+      expiresInTime,
+      entry.decodedToken.claims.exp
+    );
+
+    return {
+      body: entry,
+      expiresAt: expirySeconds
+    };
+  }
+
+  /**
+   * Finds the corresponding key in the cache based on the provided cache key.
+   * The keys inside the cache are in the format {prefix}::{client_id}::{audience}::{scope}.
+   * The first key in the cache that satisfies the following conditions is returned
+   *  - `prefix` is strict equal to Auth0's internally configured `keyPrefix`
+   *  - `client_id` is strict equal to the `cacheKey.client_id`
+   *  - `audience` is strict equal to the `cacheKey.audience`
+   *  - `scope` contains at least all the `cacheKey.scope` values
+   *  *
+   * @param cacheKey The provided cache key
+   * @param existingCacheKeys A list of existing cache keys
+   */
+  matchExistingCacheKey(cacheKey: CacheKey, existingCacheKeys: Array<string>) {
+    const { client_id, audience, scope } = cacheKey;
+
+    return existingCacheKeys.filter(key => {
```

Comment:
`Array.prototype.find` is unfortunately [not defined in IE11](https://caniuse.com/?search=array.find) so we would have to include another polyfill for this. I'll keep using filter here, but would you say more defensive code is needed here in case the array is empty? Looks like [accessing an invalid index](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#accessing_array_elements) returns `undefined`, which is the desired effect.